### DISCUSSION TITLE
Add more common bindings (or fully become Nim's LWJGL)

### DISCUSSION THREAD

**Discussion by mraxilus** [cavariux/nimgl#67] [Ideas]
As the title states, whilst OpenGL, Vulkan, GLFW, and Dear ImGUi are great to have all in one place, a few other bindings would cover all the bases of those looking for a Nim like LWJGL.

Examples of additional bindings which I would personally like to see:

- Assimp - Open Asset Importer Library ([existing binding](https://github.com/barcharcraz/nim-assimp))
- OpenAL ([existing binding]( https://github.com/treeform/openal))
- OpenCL ([existing binding](https://github.com/nim-lang/opencl))

I realize there already exists some bindings for these, however most are outdated, and/or have minimal documentation.
Having them under one convenient project such as nimgl, would be really beneficial for people looking to get started with Nim for projects such as game development.

**Comment by lmariscal** [comment]
I agree with you that OpenAL and OpenCL would be valuable additions to the existing bindings to cover both Audio and Compute. Assimp would be an interesting proposition, although a low priority in my opinion since atm you can cover this with specific made loaders.

Bringing back support for the STB header library would be nice, but making the generators for them is hard, since the headers heavily rely on macros.

Seeing the listed frameworks supported by LWJGL I also like the idea of Tiny File Dialogs ([existing bindings](https://github.com/juancarlospaco/nim-tinyfiledialogs)) and Vulkan Memory Allocator seems enticing, but it being written in C++ is a problem ([repo](https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator)). ShaderC would also work nicely with the Vulkan bindings.

What I have been personally debating as of late is the addition of OpenGLES, although we only support the Core standard of OpenGL, it already covers a lot of what EGL does in regard to Emscripten and the "modern" OpenGL stuff. The reason of adding these bindings would be to cover the integrated devices, such as phones and older GPUs which only support compatibility mode. I have to be honest here and say that I have not worked with EGL before, so I fail to see its importance, other that phones (which in itself is huge).

After my rambling, I do agree in the addition of OpenAL and OpenCL, I will see what I can do in the following weeks.

**Reply by mraxilus** [reply]
For STB I've currently been using [stb_image-Nim](https://gitlab.com/define-private-public/stb_image-Nim).

Regarding OpenAL/OpenCL, I'm unsure of the difficulty involved with adding bindings as I'm fairly new to Nim, but I'm willing to help out, if it's not too difficult.
